### ISSUE TITLE
Istio uses significantly high memory 

### ISSUE THREAD

**Issue by ramaraochavali** [istio/istio#39368]
### Bug Description

We upgraded to the recent Istio version and immediately we saw a jump of Istiod memory usage. We have increased the memory to 14G and still it is not sufficient. Ultimately it gets OOM Killed.

### Version

```prose
Master
```

**Comment by ramaraochavali** [comment]
Initial Analysis turns out that it is because of RDS Cache  enabled by`PILOT_ENABLE_RDS_CACHE`.

Here are some pprof outputs 

<img width="848" alt="Screen Shot 2022-06-08 at 1 13 04 PM" src="https://user-images.githubusercontent.com/17204904/172804864-60f04299-f1ef-441f-b7d4-9adb6c9932e8.png">
<img width="986" alt="Screen Shot 2022-06-08 at 1 13 47 PM" src="https://user-images.githubusercontent.com/17204904/172804887-cfea952c-5235-441f-b0e3-f99a25d76cb1.png">
<img width="1172" alt="Screen Shot 2022-06-08 at 1 16 43 PM" src="https://user-images.githubusercontent.com/17204904/172804892-b065138a-622a-4893-84be-acf17d69c651.png">

**Comment by hzxuzhonghu** [comment]
Config.Spec has a reference to underlying vs

**Comment by hzxuzhonghu** [comment]
And actually lru route cache key is only md5 checksum, not sure how you profile. And from your profile, I could not see how much memory totally istiod costs.

**Comment by ramaraochavali** [comment]
> Config.Spec has a reference to underlying vs

Yes

> And actually lru route cache key is only md5 checksum, not sure how you profile. And from your profile, I could not see how much memory totally istiod costs.

@deveshkandpal1224 can you please share complete profile before/after RDS cache and the memory usage?

**Comment by howardjohn** [comment]
can you upload the actual profile targz as well when you get the new one

**Comment by deveshkandpal1224** [comment]
@howardjohn @hzxuzhonghu - sorry for the delay, attached profiles as requested. PTAL - let me know if there is anything else that you need. Thanks!

low-memory -> 

[pprof.pilot-low-mem-discovery.alloc_objects.alloc_space.inuse_objects.inuse_space.005.pb.gz](https://github.com/istio/istio/files/8874457/pprof.pilot-low-mem-discovery.alloc_objects.alloc_space.inuse_objects.inuse_space.005.pb.gz)


high memory -> 

[pprof.pilot-discovery-high-mem.alloc_objects.alloc_space.inuse_objects.inuse_space.003.pb.gz](https://github.com/istio/istio/files/8874458/pprof.pilot-discovery-high-mem.alloc_objects.alloc_space.inuse_objects.inuse_space.003.pb.gz)

[pprof.pilot-discovery-high-mem.alloc_objects.alloc_space.inuse_objects.inuse_space.004.pb.gz](https://github.com/istio/istio/files/8874459/pprof.pilot-discovery-high-mem.alloc_objects.alloc_space.inuse_objects.inuse_space.004.pb.gz)

**Comment by howardjohn** [comment]
Wow so if I read this right, 4gb of keys. key is 32bytes. That is 125 000 000 keys. But XDS cache size is 60000 max. So how can we have so much memory to store the key...?

**Comment by howardjohn** [comment]
Also seems like we _actively_ have 626mb in the `params = append(params, vs.Name+"/"+vs.Namespace)` code of Key(). That doesn't seem right, it should free immediately - maybe just GC delay though. I would note we could trivially optimize that to call `hash.Write` immediately rather than have an intermediate

**Comment by ramaraochavali** [comment]
> Wow so if I read this right, 4gb of keys. key is 32bytes. That is 125 000 000 keys. But XDS cache size is 60000 max. So how can we have so much memory to store the key...?

Is some thing wrong with our key calculation or is each key is unnecessarily holding on to some stuff?

**Comment by sschepens** [comment]
@howardjohn i don't really know a lot of Istio code, but it does seem that we're leaking things in `XdsCache`, specifically in `configIndex`.

When an entry gets added we "update" the index [here](https://github.com/istio/istio/blob/master/pilot/pkg/model/xds_cache.go#L242). This ends up adding entries to `configIndex` but don't seem to ever be cleared. There is a `Clear` method on the `XdsCache` but it is only called from one place, so i'm guessing we're leaking things here.
What happens if the internal LRU evicts an entry? it would seem we definitely leak things in this case.
Is it correct for us to be wrapping LRU this way?

**Comment by howardjohn** [comment]
I haven't had a chance to dig into it but it seems plausible that would cause it

**Comment by hzxuzhonghu** [comment]
![perf](https://user-images.githubusercontent.com/13374016/173268751-e3b2bd94-4d49-47e9-bffc-ff8caa1a8ec3.PNG)

How many resources do you create? I guess it is the scalability test.

**Comment by ramaraochavali** [comment]
Not really.  It is one of the biggest clusters. But other clusters also have seen increase in memory but not at this level.

**Comment by sschepens** [comment]
@howardjohn another thing to note of the current cache implementation, the internal LRU store is bounded with a max size, but everytime we add an entry we're adding stuff to `configIndex` and `typeIndex` in an **unbounded** fashion, we're not removing elements from these maps when LRU evicts an entry, so these maps can grow unbounded.

I still don't know if we actually leak things or not, but we definitely keep elements there longer than they should and we don't have mechanisms to limit the growth of those indexes.

I also don't find a way to easily fix this, since the evict function receives the `key` and `value` that we store in the LRU store but we cannot derive the keys of `indexConfig` and `typeConfig` from this keys or values.

**Comment by sschepens** [comment]
@howardjohn @ramaraochavali @hzxuzhonghu what about something like this? https://github.com/istio/istio/pull/39435

I don't know if this would be the right approach, though.
We would have some increased memory overhead of storing the dependencies (might not be too much) and writes that cause evictions would be slower because we now have to clear the evicted key.

**Comment by sschepens** [comment]
@ramaraochavali do you have the cache metrics enabled? is it possible for you to check if there were evictions on the cluster which showed the huge memory increase?

If there are no evictions, then my theory would be incorrect.

**Comment by hzxuzhonghu** [comment]
@sschepens No, your theory is correct forever, regardless of the issue. We should think about how to fix it, but not to introduce more mem cost

**Comment by ramaraochavali** [comment]
@sschepens yes we have and it seems there are evictions - Check the last one month data
<img width="1387" alt="Screenshot 2022-06-14 at 12 38 24 PM" src="https://user-images.githubusercontent.com/17204904/173515340-052e5fc9-7fa7-450f-a9c7-c50fae3acbd4.png">

**Comment by ramaraochavali** [comment]
> Is there any chance you could deploy my PR to one of those clusters and get another profile?

It is little tough. But will try and let you know.

**Comment by nirvanagit** [comment]
@ramaraochavali: is memory leak mentioned in this issue applicable to 1.13 as well? I see from the issue description, that you did an upgrade to the master branch.

**Comment by ramaraochavali** [comment]
@nirvanagit It should be applicable I guess. But the impact may vary as the XdsCache is there in 1.13

**Comment by ramaraochavali** [comment]
@sschepens So we have deployed the fix and enabled RDS cache back. We still see Istiod getting OOM Killed. Here are the latest pprof outputs
[pprof.pilot-discovery.alloc_objects.alloc_space.inuse_objects.inuse_space.005.pb.gz](https://github.com/istio/istio/files/8989979/pprof.pilot-discovery.alloc_objects.alloc_space.inuse_objects.inuse_space.005.pb.gz)
[pprof.pilot-discovery.alloc_objects.alloc_space.inuse_objects.inuse_space.006.pb.gz](https://github.com/istio/istio/files/8989981/pprof.pilot-discovery.alloc_objects.alloc_space.inuse_objects.inuse_space.006.pb.gz)

**Comment by hzxuzhonghu** [comment]
cacheValue now records dependentConfigs, which makes up 52.66% memory?

**Comment by ramaraochavali** [comment]
I think the problem is this https://github.com/golang/go/issues/20135 - we need to periodically cleanup the index config map. I am looking at it

**Comment by hzxuzhonghu** [comment]
But From the result, it is not map. It is `dependentConfigs` slice

**Comment by ramaraochavali** [comment]
dependentConfigs is a set (map) not slice.

**Comment by hzxuzhonghu** [comment]
?? `DependentConfigs() []ConfigKey`

**Comment by ramaraochavali** [comment]
https://github.com/istio/istio/blob/5e7829ca6093b3fc58f38381b89f10fb13f2d4fe/pilot/pkg/model/xds_cache.go#L191 - i am suspecting this because DependentConfigs get added here

**Comment by ramaraochavali** [comment]
But the go issue talks about continuously shrinking map not the kind of usage we have. So may be that is not relevant here.

**Comment by sschepens** [comment]
@ramaraochavali is the memory growing continously? I mean, Istiod may be using more memory than before because of the route cache, but it might not be growing unbounded like before.

**Comment by sschepens** [comment]
from what I can see in the first pprof output you seem to have a lot of memory allocated on [DelegateVirtualServicesConfigKey](https://github.com/istio/istio/blob/master/pilot/pkg/model/push_context.go#L893) which would mean you have Virtual Services with many delegates, this ends up as part of [DependentConfigs](https://github.com/istio/istio/blob/master/pilot/pkg/networking/core/v1alpha3/route/route_cache.go#L93) so it's keeping all those references in memory.

The second pprof recording is different, there seems to be little DelegateVirtualServices but a lot of `DependentConfigs`, so maybe just a lot of stuff is getting saved in the cache.

Maybe reducing the cache size would help?

**Comment by ramaraochavali** [comment]
> from what I can see in the first pprof output you seem to have a lot of memory allocated on [DelegateVirtualServicesConfigKey](https://github.com/istio/istio/blob/master/pilot/pkg/model/push_context.go#L893) which would mean you have Virtual Services with many delegates, this ends up as part of [DependentConfigs](https://github.com/istio/istio/blob/master/pilot/pkg/networking/core/v1alpha3/route/route_cache.go#L93) so it's keeping all those references in memory.

Yes we have more delegates. This is what I also suspect.

> @ramaraochavali is the memory growing continously? I mean, Istiod may be using more memory than before because of the route cache, but it might not be growing unbounded like before.

It is growing continuously and at some points get OOM killed. If we turn of RDS cache the issue goes away.

Not sure if reducing cache size will help - We still enable CDS cache but do not see issues.

I suspect some thing wrong with `DelegateVirtualServicesConfigKey` - but not sure what it is. 

```
3.47GB 52.66% 52.66%     3.47GB 52.66%  istio.io/istio/pilot/pkg/networking/core/v1alpha3/route.(*Cache).DependentConfigs
    1.88GB 28.55% 81.21%     1.88GB 28.55%  istio.io/istio/pkg/util/sets.Set.Insert
    0.39GB  5.92% 87.12%     0.39GB  5.92%  istio.io/istio/pilot/pkg/model.(*PushContext).DelegateVirtualServicesConfigKey
    0.12GB  1.83% 88.96%     0.12GB  1.83%  istio.io/istio/pilot/pkg/model.SelectVirtualServices.func1
    0.11GB  1.74% 90.69%     0.11GB  1.74%  istio.io/pkg/ledger.(*smt).loadBatch
    0.11GB  1.62% 92.32%     0.11GB  1.62%  reflect.New
    0.08GB  1.14% 93.46%     0.08GB  1.14%  istio.io/istio/pilot/pkg/networking/core/v1alpha3/route.(*Cache).Key
    0.06GB  0.87% 94.32%     0.06GB  0.87%  istio.io/istio/pilot/pkg/model.(*PushContext).VirtualServicesForGateway
    0.04GB  0.67% 95.00%     0.04GB  0.67%  reflect.mapassign_faststr
    0.02GB  0.23% 95.23%     0.13GB  1.96%  istio.io/pkg/ledger.(*smt).loadChildren
    
```

**Comment by howardjohn** [comment]
[]model.ConfigKey from DependantConfigs is the bulk. So a bit list of ConfigKey.

We have a constant problem of ConfigKey taking too much space. Maybe time to microptimize the struct itself.

For example, if we look at 3 variants:

```

func BenchmarkConfigKey(b *testing.B) {
	b.ReportAllocs()
	names := []string{}
	for i := 0; i < 100; i++ {
		names = append(names, "my-config-", fmt.Sprint(i))
	}
	joined := []string{}
	for i := 0; i < 100; i++ {
		joined = append(joined, names[i]+"/"+names[i])
	}
	b.Run("old", func(b *testing.B) {
		for n := 0; n < b.N; n++ {
			ck := []ConfigKey{}
			for i := 0; i < 100; i++ {
				ck = append(ck, ConfigKey{
					Kind:      gvk.DestinationRule,
					Name:      names[i],
					Namespace: names[i],
				})
			}
		}
	})
	b.Run("2", func(b *testing.B) {
		for n := 0; n < b.N; n++ {
			ck := []ConfigKey2{}
			for i := 0; i < 100; i++ {
				ck = append(ck, ConfigKey2{
					Kind:      Bar,
					Name:      names[i],
					Namespace: names[i],
				})
			}
		}
	})
	b.Run("3", func(b *testing.B) {
		for n := 0; n < b.N; n++ {
			ck := []ConfigKey3{}
			for i := 0; i < 100; i++ {
				ck = append(ck, ConfigKey3{
					Kind:          Bar,
					NameNamespace: joined[i],
				})
			}
		}
	})
}

type GVK uint8

const (
	Bar GVK = iota
)

// ConfigKey describe a specific config item.
// In most cases, the name is the config's name. However, for ServiceEntry it is service's FQDN.
type ConfigKey struct {
	Kind      config.GroupVersionKind
	Name      string
	Namespace string
}
type ConfigKey2 struct {
	Kind      GVK
	Name      string
	Namespace string
}
type ConfigKey3 struct {
	Kind          GVK
	NameNamespace string
}
```

And compare allocating 100 of them into a list:
```
BenchmarkConfigKey
BenchmarkConfigKey/old
BenchmarkConfigKey/old-48                 169239              7095 ns/op           21424 B/op          8 allocs/op
BenchmarkConfigKey/2
BenchmarkConfigKey/2-48                   287408              4183 ns/op           10592 B/op          8 allocs/op
BenchmarkConfigKey/3
BenchmarkConfigKey/3-48                   541184              2223 ns/op            6120 B/op          8 allocs/op
```

unsafe.SizeOf shows essentially the same, unsurprisingly: `80, 40, 24` respectively.

The string is heap allocated though, so we should only be copying around copies for namespace/name. For the most part at least - we aren't doing anything super efficient like de-duping `namespace` references. But we have a single Config that ends up making 1000s of ConfigKeys which should all use the same shared string backing, so I think its mostly the pointers that are expensive.

**Comment by howardjohn** [comment]
RDS is unique since dependantConfigs is massive. For CDS its only O(2+EnvoyFilters). Envoyfilter count is typically low.

Wonder if we would be better off without the RDS cache until we switch to VHDS.. or we could make it not cache if we have too many dependent configs or something

**Comment by howardjohn** [comment]
We could also consider using pointers to ConfigKeys and generate a single ConfigKey per `Config` object somehow which we pass around...

**Comment by sschepens** [comment]
@howardjohn the `ConfigKey3` benchmarks are not exactly correct, since you're joining the strings outside the benchmark, in real world, we would have to join them every time we create a `ConfigKey`, that would also mean that we have to copy the strings to be able to join them and this would probably be costly in the long run.

Here are the results if we move the joining of strings inside the benchmark:
```
BenchmarkConfigKey/old-8         	  264463	      4464 ns/op	   21424 B/op	       8 allocs/op
BenchmarkConfigKey/2-8           	  537402	      2233 ns/op	   10592 B/op	       8 allocs/op
BenchmarkConfigKey/3-8           	  298119	      3898 ns/op	    7565 B/op	     108 allocs/op
```

I was thinking the same thing about using an enum to represent the Kind, it seems to reduce memory usage by half (not taking into account the strings memory) and operations are twice as fast. Mind if I have a go at implementing this?

**Comment by howardjohn** [comment]
Hm, I thought after that I moved to joining inside and it was still faster. I guess I was also allocating the name/namespace in those tests, but that is not correct. Doing https://github.com/istio/istio/issues/39368#issuecomment-1167572505 would be amazing but seems painful to implement.  K

>  I was thinking the same thing about using an enum to represent the Kind, it seems to reduce memory usage by half (not taking into account the strings memory) and operations are twice as fast. Mind if I have a go at implementing this?

I would be in favor, my only concern is if there are places where we use kinds outside of our known set. I don't _think_ that would occur though... but not certain. Would be a shame to spend a bunch of time getting things compiling to find that at the last bit

**Comment by sschepens** [comment]
@ramaraochavali one thing to note, when we create virtual service delegates `ConfigKey`s we're not deduping anything, we're storing all delegates in an array from multiple virtual services and there might be duplicated delegates, there could be some savings in deduping these.

**Comment by sschepens** [comment]
> I would be in favor, my only concern is if there are places where we use kinds outside of our known set. I don't think that would occur though... but not certain. Would be a shame to spend a bunch of time getting things compiling to find that at the last bit

@howardjohn I think I got this compiling https://github.com/istio/istio/pull/39653 , how do you think we can check if we have other uses for those kinds?

**Comment by ramaraochavali** [comment]
> Wonder if we would be better off without the RDS cache until we switch to VHDS.. or we could make it not cache if we have too many dependent configs or something

I was thinking about this and value of RDS cache if we have too many dependent configs - Earlier we were not seeing so many evictions , but after we enabled RDS cache we started seeing lot of evictions. Probably an indication of it not being so useful. It would have been nice we have hit count metric per type - then we could have figured that out easily

**Comment by hzxuzhonghu** [comment]
@sschepens why would we cache  `dependent config`, since it does not mitigate this issue, and I think compared to not clear `configIndex`, this is more expensive. Not clear does not make much harm

We can compare the two  pprof results, little different. The second one indicates it is `dependentConfigs` that is costly. But the first time, it is not.

@ramaraochavali We can't evaluate without correct numbers of your mesh configurations/service.

**Comment by howardjohn** [comment]
Not clearing configIndex is IMO unacceptable - it is a clear memory leak

**Comment by hzxuzhonghu** [comment]
Not much, configIndex will be cleaned when a config is deleted. This could be called accumulate rather than real `leak`

**Comment by howardjohn** [comment]
Ah good point. It can still leak forever if we keep generating distinct cacheKeys without removing the config though

**Comment by hzxuzhonghu** [comment]
If `ClearAll` can be triggered, that would be ok

**Comment by ramaraochavali** [comment]
> @ramaraochavali We can't evaluate without correct numbers of your mesh configurations/service.

I agree that is what I was saying  we need to metrics to see what is going on. I don't think it is specific to our service/config though - if you have a large cluster you will run in to this. We have seen smaller clusters does not hit by this much,.

**Comment by hzxuzhonghu** [comment]
Yep, we do want to know the numbers, so we can prevent this from happen

**Comment by sschepens** [comment]
@howardjohn @ramaraochavali @hzxuzhonghu I wonder if we shouldn't have a separate cache for RDS, it seems to have a very specific and probably problematic use case, we could size and tune that cache differently.

**Comment by hzxuzhonghu** [comment]
It is not related to rds, the root cause is from xds cache. If we do not resolve it, any xds may meet this in future

**Comment by howardjohn** [comment]
RDS is the only one that can have so many dependencies. We have had a
number of customers (besides Rama here) that saw the same issue with RDS; I
haven't seen any issues with XDS.

The difference is routes are a huge resource. I am leaning towards
disabling route cache by default until we move to vhds personally.

**Comment by hzxuzhonghu** [comment]
> We have had a number of customers (besides Rama here) 

Any other issues?

From the performance view, the more dependencies one resource have, it means istio will do more iterations to build it. I cannot believe the configkeys can cost so much.

**Comment by sschepens** [comment]
I think I have found another possible leak in XdsCache, gonna make a PR for fixing it and try to add some tests to validate that we don't leak these again.

**Comment by ramaraochavali** [comment]
@sschepens great. Curious what is it?

**Comment by sschepens** [comment]
Hmm i think it was not actually a leak, When we call `cache.Clear()` we remove items from cache which are dependent on the keys we provide, but we weren't explicitly cleaning the indexes.

It's not actually a leak because LRU calls evict callback whenever an entry is removed from the cache, so with the change I made previously we're now correctly clearing the indexes on `Clear`

I'm still gonna open a PR with a few tests that verify this.

**Comment by sschepens** [comment]
@ramaraochavali would it be possible to test https://github.com/istio/istio/pull/39653 now that it's merged and see how much of a difference it makes on your workloads?

**Comment by ramaraochavali** [comment]
So we have run with the latest master and enabled `RDS_CACHE` and this is what we found

- RDS has too many dependent configs.  Comparison with CDS cache vs. RDS cache dependent configs is 140000 vs. 
95421019 i.e 68057 times more . So I am not sure if any amount of optimization can help here.

<img width="1357" alt="Screenshot 2022-07-09 at 12 48 04 PM" src="https://user-images.githubusercontent.com/17204904/178230181-5bf07686-8dc7-4d96-82e3-8711572586cb.png">
<img width="1401" alt="Screenshot 2022-07-09 at 12 48 56 PM" src="https://user-images.githubusercontent.com/17204904/178230214-5267b314-dcd9-46e0-a63a-51c98769f689.png">

-  With recent fixes we made the overall istiod memory usage less - the spike lines are when RDS cache is enabled. Right data is after fixes and left to that is before the fixes.

<img width="1390" alt="Screenshot 2022-07-11 at 12 22 56 PM" src="https://user-images.githubusercontent.com/17204904/178230841-4e1b8a25-9588-49f8-8c6b-09526e1b7da2.png">

-  Still there is leak in Dependent Configs. They keep growing
<img width="1401" alt="Screenshot 2022-07-11 at 12 23 34 PM" src="https://user-images.githubusercontent.com/17204904/178231142-c1594320-39fe-4c9f-a77b-f3121071a0a4.png">

- The xds evictions metric is little unclear. The evict gets incremented even when clear is called.

**Comment by sschepens** [comment]
@ramaraochavali thanks for the update, I was wondering what the status was with the changes that have been happening, did you manage to test the fix you merged the other day?

**Comment by ramaraochavali** [comment]
>  I was wondering what the status was with the changes that have been happening, did you manage to test the fix you merged the other day?

We are planning to deploy it. Will provide updates soon.

**Comment by ramaraochavali** [comment]
Looks like the dependent config leak is fixed
<img width="1362" alt="Screenshot 2022-07-16 at 10 31 52 AM" src="https://user-images.githubusercontent.com/17204904/179341215-b7756cd0-43cd-424b-a823-516c3418860a.png">

The only remaining issue is what do we do with RDS_CACHE - Should we default it to false or leave it as is?
@howardjohn @hzxuzhonghu thoughts?

**Comment by sschepens** [comment]
@ramaraochavali what's the memory usage now?
How is CPU usage compared to before?
It would be nice to see what kind of trade offs we are making and if it's actually any use.

**Comment by ramaraochavali** [comment]
CPU Profile:

<img width="1409" alt="Screenshot 2022-07-18 at 2 00 12 PM" src="https://user-images.githubusercontent.com/17204904/179473313-9a0309ee-8f10-44f7-8e09-aa85728236c7.png">

Memory:

<img width="1360" alt="Screenshot 2022-07-18 at 11 02 45 AM" src="https://user-images.githubusercontent.com/17204904/179473335-c1cadb31-3b93-4b27-9920-020885c840e0.png">

@sschepens ^^. I could not find any thing alarming

**Comment by ramaraochavali** [comment]
it is container_cpu_usage_total_in_seconds - cadvisor metric

**Comment by sschepens** [comment]
Hmm i'm not sure I understand the CPU metric @ramaraochavali is this cumulative cpu seconds? That would not be very comparable between two deployments.
I would rather have average and probably max cpu utilization of istiod pods.

What I'm meaning to say is, it would be great to compare how the CPU usage is when enabling RDS_CACHE in the latest versions. Since we're basically using caches to save computing power, i would expect memory usage to be higher, but also cpu usage to be considerably lower, if the second doesn't hold, then caching is not working for your use case.

What I also don't understand is when the rollout of the new Istio version is done.

Is it too much to ask to provide two separate graphs for each metric? one for memory usage of the previous version and one for the latest, one for avg cpu usage of previous version and one for the latest. It would also probably be fair to compare on similar situations: aproximate load, connected sidecars, etc.

**Comment by ramaraochavali** [comment]
@sschepens Let me collect the data - but it will take some time.

**Comment by obailiumingo** [comment]
I got an error that is same to you.  istio version 1.6.3. the memory of istiod is over 50G.
when i read the code for two days and got nothing, i noticed that 1.6.3 is not the release version. so i compile the source of 1.6, then the problem fixed. mayby you should use the release version but not master.

**Comment by BigYao-c** [comment]
Hi,There. I met the same problem in version 1.13.3. Can anyone please cherry pick the fix to 1.13? Thanks a lot.

**Comment by ramaraochavali** [comment]
Can you please try setting `PILOT_ENABLE_RDS_CACHE` to false and if it helps?

**Comment by hzxuzhonghu** [comment]
> Hi,There. I met the same problem in version 1.13.3. Can anyone please cherry pick the fix to 1.13? Thanks a lot.

We need to backport the fix to 1.13 and also 1.14

**Comment by senior88oqz** [comment]
> We need to backport the fix to 1.13 and also 1.14

@hzxuzhonghu i wondered if this is affecting all the versions of 1.13/1.14?

**Comment by hzxuzhonghu** [comment]
Yes
